scripts/lib.sh

```diff
@@ -251,22 +273,63 @@ generate_dockerfile()
 RUN tar -C /usr/ -xzf /tmp/go${GO_VERSION}.linux-${goarch}.tar.gz
 ENV GOROOT=/usr/go
 ENV PATH=\$PATH:\$GOROOT/bin:\$GOPATH/bin
 "
 
-	readonly dockerfile_template="Dockerfile.in"
+	# Rust agent
+	# rust installer should set path apropiately, just in case
+	local cmake_file="cmake-${CMAKE_VERSION}.tar.gz"
+	local cmake_dir="cmake-${CMAKE_VERSION}"
+	readonly install_cmake="
+RUN pushd /root; wget https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION}/${cmake_file}; tar -zxf ${cmake_file}; cd ${cmake_dir}; ./bootstrap; make; make install; popd
+"
+	local musl_tar="musl-${MUSL_VERSION}.tar.gz"
+	local musl_dir="musl-${MUSL_VERSION}"
+	readonly install_musl="
+RUN pushd /root; wget -c https://www.musl-libc.org/releases/${musl_tar}; tar -zxf ${musl_tar}; cd ${musl_dir}; sed -i \"s/^ARCH = .*/ARCH = ${muslarch}/g\" dist/config.mak; ./configure; make; make install; echo \"/usr/local/musl/lib\" > /etc/ld-musl-${muslarch}.path; popd
+ENV PATH=\$PATH:/usr/local/musl/bin
+"
+	readonly install_rust="
+RUN curl https://sh.rustup.rs -sSf --output /tmp/rust-init; chmod a+x /tmp/rust-init; export http_proxy=${http_proxy:-}; export https_proxy=${http_proxy:-}; /tmp/rust-init -y
+RUN . /root/.cargo/env
+ENV PATH=\$PATH:/root/.cargo/bin
+RUN . /root/.cargo/env; export http_proxy=${http_proxy:-}; export https_proxy=${http_proxy:-}; cargo install cargo-when; rustup toolchain install ${RUST_VERSION}; rustup default ${RUST_VERSION}; rustup target install ${rustarch}-unknown-linux-musl
+RUN ln -sf /usr/bin/g++ /bin/musl-g++
+"
+	# rust agent still need go to build
+	# because grpc-sys need go to build
 	pushd ${dir}
 	[ -f "${dockerfile_template}" ] || die "${dockerfile_template}: file not found"
-	sed \
-		-e "s|@GO_VERSION@|${GO_VERSION}|g" \
-		-e "s|@OS_VERSION@|${OS_VERSION:-}|g" \
-		-e "s|@INSTALL_GO@|${install_go//$'\n'/\\n}|g" \
-		-e "s|@SET_PROXY@|${set_proxy:-}|g" \
-		${dockerfile_template} > Dockerfile
+	# powerpc have no musl target, don't setup rust enviroment
+	# since we cannot static link agent. Besides, there is
+	# also long double representation problem when building musl-libc
+	if [ "${architecture}" == "ppc64le" ]; then
+		sed \
+			-e "s|@GO_VERSION@|${GO_VERSION}|g" \
+			-e "s|@OS_VERSION@|${OS_VERSION:-}|g" \
+			-e "s|@INSTALL_CMAKE@||g" \
+			-e "s|@INSTALL_MUSL@||g" \
+			-e "s|@INSTALL_GO@|${install_go//$'\n'/\\n}|g" \
+			-e "s|@INSTALL_RUST@||g" \
+			-e "s|@SET_PROXY@|${set_proxy:-}|g" \
+			${dockerfile_template} > Dockerfile
+	else
+		sed \
+			-e "s|@GO_VERSION@|${GO_VERSION}|g" \
+			-e "s|@OS_VERSION@|${OS_VERSION:-}|g" \
+			-e "s|@INSTALL_CMAKE@|${install_cmake//$'\n'/\\n}|g" \
+			-e "s|@INSTALL_MUSL@|${install_musl//$'\n'/\\n}|g" \
+			-e "s|@INSTALL_GO@|${install_go//$'\n'/\\n}|g" \
+			-e "s|@INSTALL_RUST@|${install_rust//$'\n'/\\n}|g" \
+			-e "s|@SET_PROXY@|${set_proxy:-}|g" \
+			${dockerfile_template} > Dockerfile
```

Comment:
These two block of code are almost identical, but you need to look carefully to see that. I suggest you restructure the code to make the intent clearer:

```sh
sed \
    -e "s|@GO_VERSION@|${GO_VERSION}|g" \
      :
    -e "s|@SET_PROXY@|${set_proxy:-}|g" \
    ${dockerfile_template} > Dockerfile

[ "${architecture}" != "ppc64le" ] && sed -i -e "s|@INSTALL_RUST@|${install_rust//$'\n'/\\n}|g" Dockerfile
```

Reply:
we use the same build image for both go agent and rust agent on all architecture except ppc64 . but we cannot even install rust/musl on ppc64. That's why I exclude ppc64. 

The above pattern won't work